lib/src/action.dart

```diff
@@ -0,0 +1,53 @@
+library action;
+
+import 'dart:async';
+
+import 'package:greencat/src/middleware_api.dart';
+import 'typedefs.dart';
+
+/// Actions are payloads of information that send data from your application to
+/// your store. They are the only source of information for the store. You send
+/// them to the store using `store.dispatch()`.
+///
+/// Uses ideas from https://github.com/acdlite/flux-standard-action
+abstract class Action<T> {
+  ///The optional error property MAY be set to true if the action represents an
+  /// error.
+  ///
+  /// An action whose error is true is analogous to a rejected Promise. By
+  /// convention, the payload SHOULD be an error object.
+  ///
+  /// If error has any other value besides true, including undefined and null,
+  /// the action MUST NOT be interpreted as an error.
+  @deprecated
+  bool get isError => false;
+
+  /// The optional payload property MAY be any type of value. It represents the
+  /// payload of the action. Any information about the action that is not the
+  /// type or status of the action should be part of the payload field.
+  ///
+  /// By convention, if error is true, the payload SHOULD be an error object.
+  /// This is akin to rejecting a promise with an error object.
+  // TODO: Consider using the Option or Either type to convey an error.
+  dynamic get payload;
+
+  /// The type of an action identifies to the consumer the nature of the action
+  /// that has occurred. Two actions with the same type MUST be strictly
+  /// equivalent (using ===). By convention, type is usually string constant or
+  /// a Symbol, it is recommended, to use an enumeration for this
+  /// getter.
+  T get type;
+
+  @override
+  String toString() {
+    return 'Action{ type: $type, payload: $payload }';
+  }
+}
+
+/// Base type for actions that trigger side effects.
+abstract class AsyncAction<T> extends Action<T> implements Function {
+  /// Async actions are functions that return a future, it is OK to have side
+  /// effects within the body of an async action, also a Future should be
+  /// returned that resolves when all async effects from this action are resolved.
+  Future call(MiddlewareAPI api);
```

Comment:
Done. I don't think this API is good enough, right now the result is being discarded. Created https://github.com/alexeieleusis/greencat/issues/3 to follow up.